prover/prover.go

```diff
@@ -191,47 +192,74 @@ func makeRecoveryProofTree(
 	return treeCache, tree, nil
 }
 
-func generateProof(
+func sequentialWork(
 	ctx context.Context,
 	labelHashFunc func(data []byte) []byte,
 	tree *merkle.Tree,
 	treeCache *cache.Writer,
-	end time.Time,
 	nextLeafID uint64,
 	securityParam uint8,
 	persist persistFunc,
-) (uint64, *shared.MerkleProof, error) {
-	makeLabel := shared.MakeLabelFunc()
+) (uint64, error) {
 	leaves := nextLeafID
-	for leafID := nextLeafID; time.Until(end) > 0; leafID++ {
-		// Handle persistence.
+	makeLabel := shared.MakeLabelFunc()
+	for leafID := nextLeafID; ; leafID++ {
 		select {
 		case <-ctx.Done():
 			if err := persist(ctx, tree, treeCache, leafID); err != nil {
-				return 0, nil, fmt.Errorf("%w: error happened during persisting: %v", ErrShutdownRequested, err)
+				return 0, fmt.Errorf("persisting execution state: %w", err)
 			}
-			return 0, nil, ErrShutdownRequested
+			return leaves, nil
 		default:
 		}
 
 		if leafID != 0 && leafID%hardShutdownCheckpointRate == 0 {
 			if err := persist(ctx, tree, treeCache, leafID); err != nil {
-				return 0, nil, err
+				return 0, err
 			}
 		}
 
 		// Generate the next leaf.
 		err := tree.AddLeaf(makeLabel(labelHashFunc, leafID, tree.GetParkedNodes()))
 		if err != nil {
-			return 0, nil, err
+			return 0, err
 		}
 		leaves++
 	}
+}
+
+func generateProof(
+	ctx context.Context,
+	labelHashFunc func(data []byte) []byte,
+	tree *merkle.Tree,
+	treeCache *cache.Writer,
+	end time.Time,
+	nextLeafID uint64,
+	securityParam uint8,
+	persist persistFunc,
+) (uint64, *shared.MerkleProof, error) {
+	logger := logging.FromContext(ctx)
+	logger.Info("generating proof", zap.Time("end", end), zap.Uint64("nextLeafID", nextLeafID))
+
+	poswCtx, cancel := context.WithDeadline(ctx, end)
+	defer cancel()
+	leaves, err := sequentialWork(poswCtx, labelHashFunc, tree, treeCache, nextLeafID, securityParam, persist)
+	select {
+	case <-ctx.Done():
+		if err != nil {
+			return 0, nil, fmt.Errorf("%w: %v", ErrShutdownRequested, err)
+		}
+		return 0, nil, ErrShutdownRequested
+	default:
+	}
```

Comment:
I would move that into the `sequentialWork` function; on line 212:

```golang
return leaves, ctx.Err()
```

A `context.Cancelled` error already implies that a shutdown / cancellation was requested. The dedicated type `ErrShoutdownRequested` we currently have for it could just replaced in tests with `context.Cancelled`.

Reply:
Removed `ErrShoutdownRequested`